File: day2/day2.ts
class passwordRow {
    min: number;
    max:number;
    char:string; 
    password: string;
    constructor(min: number, max:number, char:string, password: string){
        this.min = min;
        this.max = max;
        this.char = char;
        this.password = password;
    }
}

async function part1(passwordRows: Array<passwordRow>){
    const validPasswords: Array<passwordRow> = [];
    for (let i=0; i<passwordRows.length; ++i){
        const v = passwordRows[i];
        const charOcc = (v.password.match(new RegExp(v.char, "g")) || []).length;
        if ((v.min <= charOcc) && (charOcc <= v.max)) validPasswords.push(v);
    }
    console.log(validPasswords.length);
}

async function part2(passwordRows: Array<passwordRow>){
    const validPasswords: Array<passwordRow> = [];
    for (let i=0; i<passwordRows.length; ++i){
        const v = passwordRows[i];
        if ((v.password[v.min-1] == v.char && v.password[v.max-1] != v.char) || 
        (v.password[v.min-1] != v.char && v.password[v.max-1] == v.char))
            validPasswords.push(v);
    }
    console.log(validPasswords.length);
}

export async function initDay2(inputStream: string){
    const inputStreamSplitted = inputStream.trim().split("\n");
    const inputData: Array<passwordRow> = [];
    for (let i = 0; i<inputStreamSplitted.length; ++i){
        const v = inputStreamSplitted[i];
        const left = v.split(":")[0];
        const right = v.split(":")[1].trim();
        const min = parseInt(left.split("-")[0], 10);
        const max = parseInt(left.split("-")[1], 10);
        const char = left.split(" ")[1];
        const data = new passwordRow(min, max, char, right);
        inputData.push(data);
    }
    await part1(inputData);
    await part2(inputData);
}

File: day2/day2.py
class PasswordRow:
    min: int
    max: int
    char: str
    password: str

    def __init__(self, data: str):
        left = data.split(":")[0]
        right = data.split(":")[1]
        # range
        ran = left.split()[0]
        self.char = left.split()[1]
        self.min = int(ran.split("-")[0])
        self.max = int(ran.split("-")[1])
        self.password = right.strip()


# array of passwordRow
def part1(passwordRows):
    v: int = 0
    for r in passwordRows:
        if r.min <= r.password.count(r.char) <= r.max:
            v = v + 1
    return v


def part2(passwordRows):
    v: int = 0
    for r in passwordRows:
        if (
            r.password[r.min - 1] is r.char and r.password[r.max - 1] is not r.char
        ) or (r.password[r.min - 1] is not r.char and r.password[r.max - 1] is r.char):
            v += 1
    return v


def day2(data: str, printable: bool = 1):
    passwordRows = []
    for i in data:
        passwordRows.append(PasswordRow(i))
    p1, p2 = part1(passwordRows), part2(passwordRows)
    if printable:
        print(p1, p2)
    return p1, p2